File: src/network/Receiver.java
package network;
import java.io.DataInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.Socket;
import java.nio.charset.StandardCharsets;

public class Receiver extends Thread {
	Socket mSocket;
	InputStream mInputStream;
	DataInputStream mDataInputStream;
	ReceiveCallback mCallback;
	
	enum ClientType {UNSPECIFIED, CS, CPP, Java}
	ClientType mClientType = ClientType.UNSPECIFIED;
	
	public Receiver(Socket _socket, InputStream _inStream, ReceiveCallback _callback) {
		mSocket = _socket;
		mInputStream = _inStream;
		mDataInputStream = new DataInputStream(_inStream);
		mCallback = _callback;
	}
	
	public synchronized void run() {
		while(!mSocket.isClosed()) {
			try {
				String stringBuffer = "";
				int incomingSize = 0;
				
				switch (mClientType) {
				case CS:
					incomingSize = mDataInputStream.readByte();
					stringBuffer = readString(incomingSize);
					mCallback.onReceiveMessage(stringBuffer);
					break;
				
				case Java:
					String request = mDataInputStream.readUTF();
					mCallback.onReceiveMessage(request);
					break;
					
				case CPP:
					incomingSize = mDataInputStream.readByte();
					stringBuffer = readString(incomingSize);
					mCallback.onReceiveMessage(stringBuffer);
					break;
					
				case UNSPECIFIED:
					incomingSize = mDataInputStream.readByte();
					mClientType = ClientType.values()[incomingSize];
					break;
				}
			}
			catch (IOException e) {
				try {
					//System.out.println("An error occured while receiving incoming data. Type: " + e.getClass().toString());
					mSocket.close();
				} catch (IOException e1) {
					//System.out.println("An error occured while closing the socket. Type: " + e1.getClass().toString());
				}
			}
		}
	}
	
	String readString(int _size) throws IOException {
		byte[] messageBinary = new byte[_size];
		for (int i = 0; i < messageBinary.length; i++) {
			messageBinary[i] = mDataInputStream.readByte();
		}
		return new String(messageBinary, StandardCharsets.UTF_8);
	}
}

File: src/network/Receiver.kt
package network

import java.io.DataInputStream
import java.io.IOException
import java.io.InputStream
import java.net.Socket
import java.nio.charset.StandardCharsets

class Receiver(var mSocket: Socket, var mInputStream: InputStream, _callback: ReceiveCallback) : Thread() {
    var mDataInputStream: DataInputStream
    var mCallback: ReceiveCallback

    enum class ClientType {
        UNSPECIFIED, CS, CPP, Java
    }

    var mClientType = ClientType.UNSPECIFIED
    @Synchronized
    override fun run() {
        while (!mSocket.isClosed) {
            try {
                var stringBuffer = ""
                var incomingSize = 0
                when (mClientType) {
                    ClientType.CS -> {
                        incomingSize = mDataInputStream.readByte().toInt()
                        stringBuffer = readString(incomingSize)
                        mCallback.onReceiveMessage(stringBuffer)
                    }
                    ClientType.Java -> {
                        val request = mDataInputStream.readUTF()
                        mCallback.onReceiveMessage(request)
                    }
                    ClientType.CPP -> {
                        incomingSize = mDataInputStream.readByte().toInt()
                        stringBuffer = readString(incomingSize)
                        mCallback.onReceiveMessage(stringBuffer)
                    }
                    ClientType.UNSPECIFIED -> {
                        incomingSize = mDataInputStream.readByte().toInt()
                        mClientType = ClientType.values()[incomingSize]
                    }
                }
            } catch (e: IOException) {
                try {
                    //System.out.println("An error occured while receiving incoming data. Type: " + e.getClass().toString());
                    mSocket.close()
                } catch (e1: IOException) {
                    //System.out.println("An error occured while closing the socket. Type: " + e1.getClass().toString());
                }
            }
        }
    }

    @Throws(IOException::class)
    fun readString(_size: Int): String {
        val messageBinary = ByteArray(_size)
        for (i in messageBinary.indices) {
            messageBinary[i] = mDataInputStream.readByte()
        }
        println("Message received")
        return String(messageBinary, StandardCharsets.UTF_8)
    }

    init {
        mDataInputStream = DataInputStream(mInputStream)
        mCallback = _callback
    }
}